### PR TITLE
Add common abstraction for plugins

### PR DESCRIPTION
**Pull request by willyborankin**
Add common abstraction for plugins, which helps to develop new one faster

### CODE REVIEW

**Comment by ivanyu** [comment]
What do you think about changing the GCS implementation to use this new classes right away? It might highlight potential flaws in their design.

**Comment by willyborankin** [comment]
@ivanyu I've done it. just split up PRs for simplicity ;-)

**Review by ivanyu — CHANGES_REQUESTED** [review]
One small thing to change, but overall LGTM. Let's see how it works in the real usage

File: repository-commons/src/main/java/io/aiven/elasticsearch/repositories/RepositorySettingsProvider.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import org.elasticsearch.common.settings.Settings;
+
+public abstract class RepositorySettingsProvider<T> {
+
+    private volatile RepositoryStorageIOProvider<T> repositoryStorageIOProvider;
+
+    public synchronized RepositoryStorageIOProvider<T> repositoryStorageIOProvider() throws IOException {
+        if (Objects.isNull(repositoryStorageIOProvider)) {
+            throw new IOException("Settings for GC storage hasn't been set");
```
[the remark this answers is not in the document]
Comment (willyborankin) [reply repository-commons/src/main/java/io/aiven/elasticsearch/repositories/RepositorySettingsProvider.java:32]: true :-D

**Comment by willyborankin** [comment]
@ivanyu done

**Review by ivanyu — APPROVED** [review]

**Pull request merged**